lib/screens/v2/disruption_diagram/builder.ex

```diff
@@ -0,0 +1,967 @@
+    |> do_add_slots(new_num_to_add, :right_end)
+  end
+
+  defp do_add_slots(builder, num_to_add, :left_end) do
+    {stop_data, new_left_end} = Vector.pop_last!(builder.left_end)
+
+    # If we just added the last slot from the left end, all we did was move
+    # a terminal/arrow back into the main sequence.
+    # Effectively, nothing was added to the diagram.
+    new_num_to_add = if vec_size(new_left_end) > 0, do: num_to_add - 1, else: num_to_add
+
+    builder
+    |> put_in([Access.key(:left_end)], new_left_end)
+    |> update_in([Access.key(:sequence)], &Vector.prepend(&1, stop_data))
+    |> do_add_slots(new_num_to_add, :left_end)
+  end
+
+  defp serialize_sequence(%__MODULE__{} = builder) do
+    Aja.Enum.map(builder.sequence, fn
+      %ArrowSlot{} = arrow -> %{type: :arrow, label_id: arrow.label_id}
+      %StopSlot{} = stop when stop.terminal? -> %{type: :terminal, label_id: stop.id}
+      %StopSlot{} = stop -> %{label: stop.label, show_symbol: true}
+      %OmittedSlot{} = omitted -> %{label: omitted.label, show_symbol: false}
+    end)
+  end
+
+  # Re-adds each of left_end and right_end to the main sequence as either:
+  # - a terminal stop slot if the end contains 1 stop,
+  # - a destination-arrow slot if the end contains multiple stops, or
+  # - nothing if the end contains no stops.
+  defp add_back_end_slots(builder) do
+    left_end = get_end_slot(builder.metadata, builder.left_end)
+    right_end = get_end_slot(builder.metadata, builder.right_end)
+
+    %{builder | sequence: left_end +++ builder.sequence +++ right_end}
+    |> recalculate_metadata()
+  end
+
+  defp get_end_slot(_meta, vec([])), do: Vector.new()
+
+  defp get_end_slot(_meta, vec([%{terminal?: true} = stop_data])), do: Vector.new([stop_data])
+
+  defp get_end_slot(_meta, vec([%ArrowSlot{} = predefined_destination])),
+    do: Vector.new([predefined_destination])
+
+  defp get_end_slot(meta, stops) do
+    stop_ids =
+      stops
+      |> Vector.filter(&is_struct(&1, StopSlot))
+      |> MapSet.new(& &1.id)
+
+    label_id = Label.get_end_label_id(stop_ids, meta.line, meta.branch)
+
+    Vector.new([%ArrowSlot{label_id: label_id}])
+  end
+
+  # Returns a sorted list of indices of the stops that are in the alert's informed entities.
+  # For station closures, this is the stops that are bypassed.
+  # For shuttles and suspensions, this is the stops that don't have any train service
+  # *as well as* the stops at the boundary of the disruption that don't have train service in one direction.
+  defp disrupted_stop_indices(%__MODULE__{} = builder) do
+    builder.sequence
+    |> Vector.with_index()
+    |> Vector.filter(fn
+      {%StopSlot{} = stop_data, _i} -> stop_data.disrupted?
+      {_other_slot_type, _i} -> false
+    end)
+    |> Aja.Enum.map(fn {_stop_data, i} -> i end)
+  end
+
+  # The closure has highest priority, so no other overlapping region can take stops from it.
+  defp closure_indices(builder), do: closure_ideal_indices(builder)
```

Comment:
What's the purpose of this separate def of `closure_indices` since there's only one pattern here, and no transformation happens?

Reply:
I originally separated each of these `_indices` funcs into 2 layers in order to make the code more explicit during development, but at this point they can be cleaned up. Especially since I worked with Betsy to adjust things so that the ends have lowest precedence instead of current location.

`<region>_ideal_indices` gives the indices that the region would contain if none were taken by a higher-precedence region, and `<region>_indices` gives the actual indices that it gets after other regions take away their respective pieces.

Since the closure region has the highest precedence and can't have any stops taken from it, these are one in the same.

Reply:
Clarifying since it was buried in my last comment: I cleaned up this code and removed `closure_ideal_indices` and `gap_ideal_indices`.